src/java.desktop/share/classes/sun/font/FontUtilities.java

```diff
@@ -308,6 +308,77 @@ else if (code >= 0x206a && code <= 0x206f) { // directional control
+     * is_default_ignorable() in hb-unicode.hh.
+     *
+     * <p>See the original Unicode list of default-ignorable codepoints
+     * <a href="https://www.unicode.org/Public/16.0.0/ucd/DerivedCoreProperties.txt">here</a>:
+     *
+     * <pre>
+     * 00AD         # Cf       SOFT HYPHEN
+     * 034F         # Mn       COMBINING GRAPHEME JOINER
+     * 061C         # Cf       ARABIC LETTER MARK
+     * 115F..1160   # Lo   [2] HANGUL CHOSEONG FILLER..HANGUL JUNGSEONG FILLER
+     * 17B4..17B5   # Mn   [2] KHMER VOWEL INHERENT AQ..KHMER VOWEL INHERENT AA
+     * 180B..180D   # Mn   [3] MONGOLIAN FREE VARIATION SELECTOR ONE..MONGOLIAN FREE VARIATION SELECTOR THREE
+     * 180E         # Cf       MONGOLIAN VOWEL SEPARATOR
+     * 180F         # Mn       MONGOLIAN FREE VARIATION SELECTOR FOUR
+     * 200B..200F   # Cf   [5] ZERO WIDTH SPACE..RIGHT-TO-LEFT MARK
+     * 202A..202E   # Cf   [5] LEFT-TO-RIGHT EMBEDDING..RIGHT-TO-LEFT OVERRIDE
+     * 2060..2064   # Cf   [5] WORD JOINER..INVISIBLE PLUS
+     * 2065         # Cn       <reserved-2065>
+     * 2066..206F   # Cf  [10] LEFT-TO-RIGHT ISOLATE..NOMINAL DIGIT SHAPES
+     * 3164         # Lo       HANGUL FILLER
+     * FE00..FE0F   # Mn  [16] VARIATION SELECTOR-1..VARIATION SELECTOR-16
+     * FEFF         # Cf       ZERO WIDTH NO-BREAK SPACE
+     * FFA0         # Lo       HALFWIDTH HANGUL FILLER
+     * FFF0..FFF8   # Cn   [9] <reserved-FFF0>..<reserved-FFF8>
+     * 1BCA0..1BCA3 # Cf   [4] SHORTHAND FORMAT LETTER OVERLAP..SHORTHAND FORMAT UP STEP
+     * 1D173..1D17A # Cf   [8] MUSICAL SYMBOL BEGIN BEAM..MUSICAL SYMBOL END PHRASE
+     * E0000        # Cn       <reserved-E0000>
+     * E0001        # Cf       LANGUAGE TAG
+     * E0002..E001F # Cn  [30] <reserved-E0002>..<reserved-E001F>
+     * E0020..E007F # Cf  [96] TAG SPACE..CANCEL TAG
+     * E0080..E00FF # Cn [128] <reserved-E0080>..<reserved-E00FF>
+     * E0100..E01EF # Mn [240] VARIATION SELECTOR-17..VARIATION SELECTOR-256
+     * E01F0..E0FFF # Cn [3600] <reserved-E01F0>..<reserved-E0FFF>
+     * </pre>
+     */
+    public static boolean isDefaultIgnorable(int charCode) {
+        if (charCode < 0x00AD) {
+            return false;
+        }
+        int plane = charCode >> 16;
+        if (plane == 0) {
+            // basic multilingual plane (BMP)
+            int page = charCode >> 8;
+            switch (page) {
+                case 0x00: return (charCode == 0x00AD);
+                case 0x03: return (charCode == 0x034F);
+                case 0x06: return (charCode == 0x061C);
+                case 0x17: return (charCode >= 0x17B4 && charCode <= 0x17B5);
+                case 0x18: return (charCode >= 0x180B && charCode <= 0x180E);
+                case 0x20: return (charCode >= 0x200B && charCode <= 0x200F) ||
+                                  (charCode >= 0x202A && charCode <= 0x202E) ||
+                                  (charCode >= 0x2060 && charCode <= 0x206F);
```

Comment:
I just noticed this really old open bug
https://bugs.openjdk.org/browse/JDK-6562489 : Font-Renderer should ignore invisible characters \u2062 and \u2063